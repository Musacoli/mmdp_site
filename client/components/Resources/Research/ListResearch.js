import React from 'react';
import PropTypes from 'prop-types';
import { Grid } from 'semantic-ui-react';
import BlueCard from '../../common/Card/BlueCard';
import ConfirmModal from '../../common/Modal/ConfirmModal';
import InvalidPage from '../../common/InvalidPage';
import {
  ucFirstLetter,
  ARCHIVE_ACTION,
  UNARCHIVE_ACTION,
} from '../../../utils/helper';
import Search from '../../common/Search';
import AddNewButton from '../../common/AddNewButton';

const ListResearch = ({
  results,
  isOpen,
  modalAction,
  addResearchUrl,
  instanceName,
  showModal,
  handleModalToggle,
  onConfirm,
  handleSearch,
  handleSearchChange,
  loading,
}) => {
  const generateEditLink = (id) => `/resources/research/edit/${id}`;
  const researchResults = results.results;
  let modalInfo;
  if (modalAction === 'delete') {
    modalInfo = {
      header: 'Delete research',
      content: 'Are you sure you want to delete research?',
    };
  } else if (
    modalAction === ARCHIVE_ACTION ||
    modalAction === UNARCHIVE_ACTION
  ) {
    modalInfo = {
      header: `${ucFirstLetter(modalAction)} research`,
      content: `Are you sure you want to ${modalAction} research?`,
    };
  }
  return (
    <>
      <Grid>
        <Grid.Column width={12}>
          <Search
            onSearch={handleSearch}
            onChange={handleSearchChange}
            className="research-search"
            placeholder="Search research"
          />
        </Grid.Column>
        <Grid.Column width={3}>
          <AddNewButton
            url={addResearchUrl}
            text="Research"
            className="common__button bg-ugly-blue"
          />
        </Grid.Column>
      </Grid>
      <Grid relaxed>
        {researchResults && researchResults.length
          ? researchResults.map(({ _id: id, title, archived }) => {
              const archiveAction = archived
                ? UNARCHIVE_ACTION
                : ARCHIVE_ACTION;
              return (
                <Grid.Column key={id} mobile={8} tablet={8} computer={4}>
                  <BlueCard
                    title={title}
                    archived={archived}
                    onArchive={(event) =>
                      showModal(event, { id, modalAction: archiveAction })
                    }
                    onDelete={(event) =>
                      showModal(event, { id, modalAction: 'delete' })
                    }
                    editLink={generateEditLink(id)}
                  />
                </Grid.Column>
              );
            })
          : !loading && (
              <Grid.Row className="ui loading center aligned animated fadeIn">
                <InvalidPage
                  pathLabel={`Add a ${instanceName}`}
                  errorMessage={`No ${instanceName} to display`}
                  errorDescription={`Please add ${instanceName}`}
                  path={addResearchUrl || '/'}
                />
              </Grid.Row>
            )}
        <ConfirmModal
          isOpen={isOpen}
          handleModalToggle={handleModalToggle}
          onConfirm={onConfirm}
          modalInfo={modalInfo}
        />
      </Grid>
    </>
  );
};

ListResearch.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  modalAction: PropTypes.string,
  showModal: PropTypes.func.isRequired,
  handleModalToggle: PropTypes.func.isRequired,
  results: PropTypes.shape({}).isRequired,
  onConfirm: PropTypes.func.isRequired,
  handleSearchChange: PropTypes.func.isRequired,
  handleSearch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
};

export default ListResearch;
