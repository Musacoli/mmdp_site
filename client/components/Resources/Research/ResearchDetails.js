import React from 'react';
import { List } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

export const ResearchCard = ({ title, Archived, _id, onDelete, onArchive }) => {
  return (
    <div className="research__details">
      <p id="research__content">{title}</p>
      <List horizontal>
        <List.Item>
          <Link to={`/resources/research/edit/${_id}`} id="research__edit__btn">
            Edit
          </Link>
        </List.Item>
        <List.Item
          as="a"
          id="research__archive__btn"
          onClick={() => {
            const data = Archived ? { Archived: false } : { Archived: true };
            onArchive(_id, data);
          }}
        >
          {Archived ? 'UnArchive' : 'Archive'}
        </List.Item>
        <List.Item
          as="a"
          id="research__delete__btn"
          onClick={() => {
            onDelete(_id);
          }}
        >
          Delete
        </List.Item>
      </List>
    </div>
  );
};

export default ResearchCard;

ResearchCard.propTypes = {
  title: PropTypes.string.isRequired,
  Archived: PropTypes.bool.isRequired,
  _id: PropTypes.string.isRequired,
  onDelete: PropTypes.func.isRequired,
  onArchive: PropTypes.func.isRequired,
};
