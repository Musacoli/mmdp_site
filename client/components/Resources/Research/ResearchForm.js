import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import FileInput from '../../common/Form/FileInput';

const ResearchForm = ({ fileName, onChange, onSubmit, loading }) => (
  <Form onSubmit={onSubmit} loading={loading} className="main-content">
    <Form.Group widths="equal">
      <Form.Field>
        <FileInput
          name="researchFile"
          label="Upload file"
          placeholder="Select a file"
          value={fileName || ''}
          className="test"
          id="researchFile"
          onChange={onChange}
        />
      </Form.Field>
      <Form.Input
        name="title"
        className="form__text-input"
        fluid
        label="Research title"
        type="text"
        placeholder="Research carried out last year"
        onChange={onChange}
      />
    </Form.Group>
    <Form.Button type="submit" className="btn__upload u-margin-top-medium">
      Upload Research
    </Form.Button>
  </Form>
);

ResearchForm.propTypes = {
  loading: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  fileName: PropTypes.string.isRequired,
};

export default ResearchForm;
