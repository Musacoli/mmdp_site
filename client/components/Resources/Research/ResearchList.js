import React from 'react';
import { Grid } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import Details from './ResearchDetails';

export const ResearchList = ({ results }) => {
  const researchResults = results.results;
  let researchDetails;
  if (researchResults) {
    researchDetails = researchResults.map((x) => {
      return (
        <Details
          title={x.title}
          key={x._id}
          _id={x._id}
          Archived={x.Archived}
        />
      );
    });
  }
  return (
    <div>
      <Grid className="Research__list">{researchDetails}</Grid>
    </div>
  );
};

export default ResearchList;

ResearchList.propTypes = {
  // results: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  results: PropTypes.shape({}).isRequired,
};
