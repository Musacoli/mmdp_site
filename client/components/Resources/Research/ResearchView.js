import React from 'react';
import { FileInput } from '../../About/Inputs/FileInput';
import './styles.scss';

const ResearchView = ({ onChange, onSubmit, fileName }) => (
  <div id="research__view">
    <div className="research__header__container">
      <h3 id="research__header">Add Research</h3>
    </div>
    <form onSubmit={onSubmit} className="ui ">
      <div className="ui grid">
        <div className="four wide column report__file__container">
          <label htmlFor="reportFile" className="title__input">
            Upload file
          </label>
          <FileInput
            placeholder={fileName || 'select a file'}
            change={onChange}
            name="reportFile"
            id="reportFile"
            value=""
          />
        </div>
        <div className="four wide column research__input">
          <div className="research__header">
            <label htmlFor="research__title" className="title__input">
              Research title
            </label>
          </div>
          <input
            type="text"
            name="title"
            id="research__title"
            className="ui input "
            placeholder="Research carried out last year"
            onChange={onChange}
            required
          />
        </div>
      </div>
      <div className="login__container">
        <button type="submit" className="ui button " id="resources__btn">
          <span id="sign__text">Upload Research</span>
        </button>
      </div>
    </form>
  </div>
);

export default ResearchView;
