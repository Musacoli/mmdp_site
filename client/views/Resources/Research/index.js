import React from 'react';
import Template from '../../Templates';
import AddResearch from '../../../containers/Resources/Research';

const ResearchView = ({ ...props }) => (
  <Template {...props} title="Add research">
    <AddResearch {...props} />
  </Template>
);

export default ResearchView;
