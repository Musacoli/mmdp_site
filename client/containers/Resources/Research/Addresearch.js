import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ResearchForm from '../../../components/Resources/Research/ResearchForm';
import { addResearch } from '../../../store/actions/resources/research';
import { validateFields } from '../../../utils/validations';

export class AddResearch extends Component {
  state = {
    title: '',
    researchFile: '',
    fileName: '',
  };

  onChange = (e) => {
    if (e.target.name === 'researchFile' && e.target.files.length) {
      const file = e.target.files[0];
      this.setState({
        fileName: file.name || '',
        researchFile: file,
      });
    } else {
      this.setState({ [e.target.name]: e.target.value });
    }
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    const details = this.state;
    const { addResearch: addResearchAction, history } = this.props;
    const formData = new FormData();
    Object.keys(details).forEach((key) => {
      formData.append(key, details[key]);
    });
    addResearchAction({ formData, history });
  };

  render() {
    const { fileName, title } = this.state;
    const { loading } = this.props;
    return (
      <div>
        <ResearchForm
          onChange={this.onChange}
          onSubmit={this.onFormSubmit}
          fileName={fileName}
          loading={loading}
          disabled={validateFields({ title, fileName })}
        />
      </div>
    );
  }
}

AddResearch.propTypes = {
  addResearch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  history: PropTypes.shape({}).isRequired,
  research: PropTypes.shape({}).isRequired,
};

const mapStateToProps = (state) => ({
  research: state.research,
  loading: state.research.loading,
});

const mapDispatchToProps = {
  addResearch,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AddResearch);
