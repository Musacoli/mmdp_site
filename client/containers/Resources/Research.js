import React, { Component } from 'react';
import { connect } from 'react-redux';
import ResearchView from '../../components/Resources/Research/ResearchView';
import { addResearch } from '../../store/actions/resources/research';
import SideBarNav from '../Sidebar';

export class Research extends Component {
  state = {
    title: '',
    reportFile: '',
    fileName: '',
  };

  onChange = (e) => {
    if (e.target.name === 'reportFile' && e.target.files.length) {
      const file = e.target.files[0];
      this.setState({
        fileName: file.name || '',
        reportFile: file,
      });
    } else {
      this.setState({ [e.target.name]: e.target.value });
    }
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    const details = this.state;
    const formData = new FormData();
    Object.keys(details).forEach((key) => {
      formData.append(key, details[key]);
    });
    this.props.addResearch(formData);
  };

  render() {
    return (
      <div className="research__container">
        <div>
          <SideBarNav />
        </div>
        <div id="research__view">
          <ResearchView
            onChange={this.onChange}
            onSubmit={this.onFormSubmit}
            fileName={this.state.fileName}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  Research: state.response,
});

const mapDispatchToProps = {
  addResearch,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Research);
