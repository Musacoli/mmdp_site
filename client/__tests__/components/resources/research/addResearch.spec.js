import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AddResearch } from '../../../../containers/Resources/Research/Addresearch';

describe('<ResearchContainer /> ', () => {
  const addResearch = () => {};
  const props = {
    research: {
      payload: {
        status: 'success',
      },
    },
    payload: {},
    loading: true,
    history: {
      push: jest.fn(),
    },
  };
  const wrapper = mount(<AddResearch addResearch={addResearch} {...props} />);
  it('renders the container without crashing', () => {
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it('tests methods in the  container', () => {
    const instance = wrapper.instance();
    wrapper.find('button').simulate('click');
    const e = {
      preventDefault: () => {},
      target: {
        name: 'form data',
        files: ['file1'],
      },
    };

    instance.onFormSubmit(e);
    instance.onChange(e);
    e.target.name = 'researchFile';
    instance.onChange(e);
  });
});
