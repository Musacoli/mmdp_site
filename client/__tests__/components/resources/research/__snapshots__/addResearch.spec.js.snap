// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResearchContainer />  renders the container without crashing 1`] = `
<AddResearch
  addResearch={[Function]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
  loading={true}
  payload={Object {}}
  research={
    Object {
      "payload": Object {
        "status": "success",
      },
    }
  }
>
  <div>
    <ResearchForm
      disabled={true}
      fileName=""
      loading={true}
      onChange={[Function]}
      onSubmit={[Function]}
    >
      <Form
        as="form"
        id="research-main-content"
        loading={true}
        onSubmit={[Function]}
      >
        <form
          className="ui loading form"
          id="research-main-content"
          onSubmit={[Function]}
        >
          <FormGroup
            widths="equal"
          >
            <div
              className="equal width fields"
            >
              <FormField>
                <div
                  className="field"
                >
                  <FileInput
                    className="test"
                    id="researchFile"
                    label="Upload file"
                    name="researchFile"
                    onChange={[Function]}
                    placeholder="Select a file"
                    value=""
                  >
                    <div
                      className="field test"
                    >
                      <label
                        htmlFor="researchFile"
                      >
                        Upload file
                      </label>
                      <div
                        className="form__file-area"
                      >
                        <input
                          className="form__file-input"
                          id="researchFile"
                          name="researchFile"
                          onChange={[Function]}
                          type="file"
                        />
                        <span
                          className="form__custom-file-input"
                        >
                          Select a file
                        </span>
                      </div>
                    </div>
                  </FileInput>
                </div>
              </FormField>
              <FormInput
                as={[Function]}
                className="form__text-input"
                control={[Function]}
                defaultValue=""
                fluid={true}
                label="Research title"
                name="title"
                onChange={[Function]}
                placeholder="Research carried out last year"
                required={true}
                type="text"
              >
                <FormField
                  className="form__text-input"
                  control={[Function]}
                  defaultValue=""
                  fluid={true}
                  label="Research title"
                  name="title"
                  onChange={[Function]}
                  placeholder="Research carried out last year"
                  required={true}
                  type="text"
                >
                  <div
                    className="required field form__text-input"
                  >
                    <label>
                      Research title
                    </label>
                    <Input
                      defaultValue=""
                      fluid={true}
                      name="title"
                      onChange={[Function]}
                      placeholder="Research carried out last year"
                      required={true}
                      type="text"
                    >
                      <div
                        className="ui fluid input"
                      >
                        <input
                          defaultValue=""
                          name="title"
                          onChange={[Function]}
                          placeholder="Research carried out last year"
                          required={true}
                          type="text"
                        />
                      </div>
                    </Input>
                  </div>
                </FormField>
              </FormInput>
            </div>
          </FormGroup>
          <FormButton
            as={[Function]}
            className="btn__upload disable-btn-area"
            control={[Function]}
            disabled={true}
            id="research__upload__btn"
            type="submit"
          >
            <FormField
              className="btn__upload disable-btn-area"
              control={[Function]}
              disabled={true}
              id="research__upload__btn"
              type="submit"
            >
              <div
                className="disabled field btn__upload disable-btn-area"
              >
                <Button
                  as="button"
                  disabled={true}
                  id="research__upload__btn"
                  type="submit"
                >
                  <button
                    className="ui disabled button"
                    disabled={true}
                    id="research__upload__btn"
                    onClick={[Function]}
                    tabIndex={-1}
                    type="submit"
                  >
                    Upload Research
                  </button>
                </Button>
              </div>
            </FormField>
          </FormButton>
        </form>
      </Form>
    </ResearchForm>
  </div>
</AddResearch>
`;
